File: webflux-dialogue/src/main/java/com/study/webflux/rag/application/monitoring/service/MetricsRollupScheduler.java
package com.study.webflux.rag.application.monitoring.service;

import java.time.Clock;
import java.time.Instant;
import java.time.temporal.ChronoUnit;

import lombok.extern.slf4j.Slf4j;

import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

import com.study.webflux.rag.domain.monitoring.model.MetricsGranularity;
import com.study.webflux.rag.domain.monitoring.model.MetricsRollup;
import com.study.webflux.rag.domain.monitoring.model.PerformanceMetrics;
import com.study.webflux.rag.domain.monitoring.model.StagePerformanceRollup;
import com.study.webflux.rag.domain.monitoring.model.UsageAnalytics;
import com.study.webflux.rag.domain.monitoring.port.MetricsRollupRepository;
import com.study.webflux.rag.domain.monitoring.port.PerformanceMetricsRepository;
import com.study.webflux.rag.domain.monitoring.port.StagePerformanceRollupRepository;
import com.study.webflux.rag.domain.monitoring.port.UsageAnalyticsRepository;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Service
@Slf4j
public class MetricsRollupScheduler {

	private final UsageAnalyticsRepository usageAnalyticsRepository;
	private final PerformanceMetricsRepository performanceMetricsRepository;
	private final MetricsRollupRepository metricsRollupRepository;
	private final StagePerformanceRollupRepository stagePerformanceRollupRepository;
	private final Clock clock;

	public MetricsRollupScheduler(UsageAnalyticsRepository usageAnalyticsRepository,
		PerformanceMetricsRepository performanceMetricsRepository,
		MetricsRollupRepository metricsRollupRepository,
		StagePerformanceRollupRepository stagePerformanceRollupRepository,
		Clock clock) {
		this.usageAnalyticsRepository = usageAnalyticsRepository;
		this.performanceMetricsRepository = performanceMetricsRepository;
		this.metricsRollupRepository = metricsRollupRepository;
		this.stagePerformanceRollupRepository = stagePerformanceRollupRepository;
		this.clock = clock;
	}

	/**
	 * 직전 1분 버킷의 사용량/스테이지 성능 데이터를 롤업해 저장합니다.
	 */
	@Scheduled(cron = "0 * * * * *")
	public void rollupMinuteMetrics() {
		Instant bucketStart = previousMinuteBucketStart();
		Instant bucketEnd = bucketStart.plus(1, ChronoUnit.MINUTES);

		Mono.when(buildUsageRollup(bucketStart, bucketEnd),
			buildStageRollup(bucketStart, bucketEnd))
			.doOnSuccess(v -> log.debug("분 단위 롤업 완료: bucketStart={}", bucketStart))
			.doOnError(error -> log.error("분 단위 롤업 실패 bucketStart={}, 이유={}",
				bucketStart,
				error.getMessage(),
				error))
			.onErrorResume(error -> Mono.empty())
			.subscribe();
	}

	private Instant previousMinuteBucketStart() {
		return clock.instant()
			.truncatedTo(ChronoUnit.MINUTES)
			.minus(1, ChronoUnit.MINUTES);
	}

	private Mono<Void> buildUsageRollup(Instant bucketStart, Instant bucketEnd) {
		return usageAnalyticsRepository.findByTimeRange(bucketStart, bucketEnd)
			.reduce(new UsageAggregate(), UsageAggregate::add)
			.defaultIfEmpty(new UsageAggregate())
			.map(aggregate -> toMetricsRollup(bucketStart, aggregate))
			.flatMap(metricsRollupRepository::save)
			.then();
	}

	private Mono<Void> buildStageRollup(Instant bucketStart, Instant bucketEnd) {
		Flux<StagePerformanceRollup> rollups = performanceMetricsRepository
			.findByTimeRange(bucketStart, bucketEnd)
			.flatMapIterable(metrics -> metrics.stages() == null
				? java.util.List.of()
				: metrics.stages())
			.groupBy(PerformanceMetrics.StagePerformance::stageName)
			.flatMap(group -> group.reduce(new StageAggregate(group.key()), StageAggregate::add))
			.map(aggregate -> aggregate.toRollup(bucketStart));

		return stagePerformanceRollupRepository.saveAll(rollups).then();
	}

	/**
	 * 사용량 집계값을 분 단위 롤업 모델로 변환합니다.
	 */
	private MetricsRollup toMetricsRollup(Instant bucketStart, UsageAggregate aggregate) {
		double avg = aggregate.requestCount == 0
			? 0
			: (double) aggregate.totalDurationMillis / aggregate.requestCount;
		return new MetricsRollup(
			bucketStart,
			MetricsGranularity.MINUTE,
			aggregate.requestCount,
			aggregate.totalTokens,
			aggregate.totalDurationMillis,
			avg);
	}

	private static final class UsageAggregate {
		private long requestCount;
		private long totalTokens;
		private long totalDurationMillis;

		private UsageAggregate add(UsageAnalytics analytics) {
			requestCount += 1;
			if (analytics.llmUsage() != null) {
				totalTokens += analytics.llmUsage().totalTokens();
			}
			if (analytics.responseMetrics() != null) {
				totalDurationMillis += analytics.responseMetrics().totalDurationMillis();
			}
			return this;
		}
	}

	private static final class StageAggregate {
		private final String stageName;
		private long count;
		private long totalDurationMillis;

		private StageAggregate(String stageName) {
			this.stageName = stageName;
		}

		private StageAggregate add(PerformanceMetrics.StagePerformance stage) {
			count += 1;
			totalDurationMillis += stage.durationMillis();
			return this;
		}

		private StagePerformanceRollup toRollup(Instant bucketStart) {
			double avg = count == 0 ? 0 : (double) totalDurationMillis / count;
			return new StagePerformanceRollup(
				bucketStart,
				MetricsGranularity.MINUTE,
				stageName,
				count,
				totalDurationMillis,
				avg);
		}
	}
}

File: app-miyou/src/main/kotlin/com/miyou/app/application/monitoring/service/MetricsRollupScheduler.kt
package com.miyou.app.application.monitoring.service

import com.miyou.app.domain.monitoring.model.MetricsGranularity
import com.miyou.app.domain.monitoring.model.MetricsRollup
import com.miyou.app.domain.monitoring.model.PerformanceMetrics
import com.miyou.app.domain.monitoring.model.StagePerformanceRollup
import com.miyou.app.domain.monitoring.model.UsageAnalytics
import com.miyou.app.domain.monitoring.port.MetricsRollupRepository
import com.miyou.app.domain.monitoring.port.PerformanceMetricsRepository
import com.miyou.app.domain.monitoring.port.StagePerformanceRollupRepository
import com.miyou.app.domain.monitoring.port.UsageAnalyticsRepository
import org.slf4j.LoggerFactory
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.time.Clock
import java.time.Instant
import java.time.temporal.ChronoUnit

@Service
class MetricsRollupScheduler(
    private val usageAnalyticsRepository: UsageAnalyticsRepository,
    private val performanceMetricsRepository: PerformanceMetricsRepository,
    private val metricsRollupRepository: MetricsRollupRepository,
    private val stagePerformanceRollupRepository: StagePerformanceRollupRepository,
    private val clock: Clock,
) {
    private val logger = LoggerFactory.getLogger(MetricsRollupScheduler::class.java)

    @Scheduled(cron = "0 * * * * *")
    fun rollupMinuteMetrics() {
        val bucketStart = previousMinuteBucketStart()
        val bucketEnd = bucketStart.plus(1, ChronoUnit.MINUTES)

        Mono
            .zip(buildUsageRollup(bucketStart, bucketEnd), buildStageRollup(bucketStart, bucketEnd))
            .doOnSuccess {
                logger.debug("분단위 집계 완료: bucketStart={}", bucketStart)
            }.doOnError { error ->
                logger.error("분단위 집계 실패: bucketStart={}, error={}", bucketStart, error.message, error)
            }.onErrorResume { Mono.empty() }
            .subscribe()
    }

    private fun previousMinuteBucketStart(): Instant =
        clock.instant().truncatedTo(ChronoUnit.MINUTES).minus(1, ChronoUnit.MINUTES)

    private fun buildUsageRollup(
        bucketStart: Instant,
        bucketEnd: Instant,
    ): Mono<Void> =
        usageAnalyticsRepository
            .findByTimeRange(bucketStart, bucketEnd)
            .reduce(UsageAggregate(), UsageAggregate::add)
            .defaultIfEmpty(UsageAggregate())
            .map { aggregate -> toMetricsRollup(bucketStart, aggregate) }
            .flatMap(metricsRollupRepository::save)
            .then()

    private fun buildStageRollup(
        bucketStart: Instant,
        bucketEnd: Instant,
    ): Mono<Void> {
        val rollups: Flux<StagePerformanceRollup> =
            performanceMetricsRepository
                .findByTimeRange(bucketStart, bucketEnd)
                .flatMapIterable { metrics ->
                    metrics.stages ?: emptyList()
                }.groupBy { stage -> stage.stageName }
                .flatMap { group -> group.reduce(StageAggregate(group.key()), StageAggregate::add) }
                .map { aggregate -> aggregate.toRollup(bucketStart) }

        return stagePerformanceRollupRepository.saveAll(rollups).then()
    }

    private fun toMetricsRollup(
        bucketStart: Instant,
        aggregate: UsageAggregate,
    ): MetricsRollup {
        val avg =
            if (aggregate.requestCount == 0L) {
                0.0
            } else {
                aggregate.totalDurationMillis.toDouble() / aggregate.requestCount
            }
        return MetricsRollup(
            bucketStart,
            MetricsGranularity.MINUTE,
            aggregate.requestCount,
            aggregate.totalTokens,
            aggregate.totalDurationMillis,
            avg,
        )
    }

    private class UsageAggregate {
        var requestCount: Long = 0
        var totalTokens: Long = 0
        var totalDurationMillis: Long = 0

        fun add(analytics: UsageAnalytics): UsageAggregate {
            requestCount += 1
            analytics.llmUsage?.let { totalTokens += it.totalTokens }
            analytics.responseMetrics?.let { totalDurationMillis += it.totalDurationMillis }
            return this
        }
    }

    private class StageAggregate(
        private val stageName: String,
    ) {
        var count: Long = 0
        var totalDurationMillis: Long = 0

        fun add(stage: PerformanceMetrics.StagePerformance): StageAggregate {
            count += 1
            totalDurationMillis += stage.durationMillis
            return this
        }

        fun toRollup(bucketStart: Instant): StagePerformanceRollup {
            val avg = if (count == 0L) 0.0 else totalDurationMillis.toDouble() / count
            return StagePerformanceRollup(
                bucketStart,
                MetricsGranularity.MINUTE,
                stageName,
                count,
                totalDurationMillis,
                avg,
            )
        }
    }
}
